### PR TITLE
feat(pubsub): examples to mock the clients

### LINKED ISSUE #4620: Implement an example showing how to mock the Cloud Pub/Sub C++ library
Also create a pubsub-mocking.dox page (look at spanner for inspiration) and link that from the landing page.

### PR DESCRIPTION
**Pull request by coryan**
This adds examples showing application developers how to mock the
`pubsub::Subscriber` and `pubsub::Publisher` clients.

Fixes #4620

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4967)

### CODE REVIEW

File: google/cloud/pubsub/doc/publisher-mock.dox [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+### Full Listing
+
+Finally we present the full code for this example:
+
+@snippet mock_publisher.cc all
+
+[googletest-link]: https://github.com/google/googletest
+[googlemock-link]: https://github.com/google/googletest/tree/master/googlemock
+
+*/
```
Comment (devbww) [inline google/cloud/pubsub/doc/publisher-mock.dox:51]: No newline at end of file.

File: google/cloud/pubsub/samples/mock_publisher.cc [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  //! [expected-results]
+  EXPECT_TRUE(id.ok());
+  EXPECT_EQ("test-id-0", *id);
+  //! [expected-results]
+}
+
+}  // namespace
+
+//! [all]
```
Comment (devbww) [inline google/cloud/pubsub/samples/mock_publisher.cc:61]: No newline at end of file.

File: google/cloud/pubsub/samples/mock_subscriber.cc [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  //! [expected-results]
+  EXPECT_TRUE(session.get().ok());
+  EXPECT_THAT(payloads,
+              UnorderedElementsAre("message-0", "message-1", "message-2"));
+  //! [expected-results]
+}
+
+}  // namespace
+
+//! [all]
```
Comment (devbww) [inline google/cloud/pubsub/samples/mock_subscriber.cc:102]: No newline at end of file.

File: google/cloud/pubsub/doc/publisher-mock.dox (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/doc/publisher-mock.dox]: Fixed (same for the others). Sorry about that.

**Pull request merged**